cpp/src/arrow/ipc/adapter.cc

```diff
@@ -0,0 +1,296 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include "arrow/ipc/adapter.h"
+
+#include <cstdint>
+#include <cstring>
+#include <vector>
+
+#include "arrow/array.h"
+#include "arrow/ipc/memory.h"
+#include "arrow/ipc/Message_generated.h"
+#include "arrow/ipc/metadata.h"
+#include "arrow/ipc/metadata-internal.h"
+#include "arrow/schema.h"
+#include "arrow/table.h"
+#include "arrow/type.h"
+#include "arrow/types/construct.h"
+#include "arrow/types/primitive.h"
+#include "arrow/util/buffer.h"
+#include "arrow/util/status.h"
+
+namespace arrow {
+namespace ipc {
+
+static bool IsPrimitive(const DataType* type) {
```

Comment:
general comment on the file, once we have a DCHECK MACRO, there are a lot of places in this file that could use it =)